cql3/restrictions/statement_restrictions.cc

```diff
@@ -1043,6 +1047,381 @@ expression make_conjunction(const expression& a, const expression& b) {
+            auto found = std::find_if(range.begin(), range.end(), [&] (auto&& element) {
+                return element_type->compare(element.serialize_nonnull(), val) == 0;
+            });
+            return found != range.end();
+        };
+        if (col_type->is_list()) {
+            if (!exists_in(value_cast<list_type_impl::native_type>(collection))) {
+                return false;
+            }
+        } else if (col_type->is_set()) {
+            if (!exists_in(value_cast<set_type_impl::native_type>(collection))) {
+                return false;
+            }
+        } else {
+            auto data_map = value_cast<map_type_impl::native_type>(collection);
+            using entry = std::pair<data_value, data_value>;
+            if (!exists_in(data_map | boost::adaptors::transformed([] (const entry& e) { return e.second; }))) {
+                return false;
+            }
+        }
+        return true;
+    });
+}
+
+/// True iff columns is a single collection containing value.
+bool contains(const raw_value_view& value, const std::vector<column_value>& columns, row_data data) {
+    if (columns.size() != 1) {
+        throw exceptions::unsupported_operation_exception("tuple CONTAINS not allowed");
+    }
+    if (columns[0].sub) {
+        throw exceptions::unsupported_operation_exception("CONTAINS lhs is subscripted");
+    }
+    const auto collection = get_value(columns[0], data);
+    if (collection) {
+        return contains(columns[0].col->type->deserialize(*collection), value);
+    } else {
+        return false;
+    }
+}
+
+/// True iff \p columns has a single element that's a map containing \p key.
+bool contains_key(const std::vector<column_value>& columns, cql3::raw_value_view key, row_data data) {
+    if (columns.size() != 1) {
+        throw exceptions::unsupported_operation_exception("CONTAINS KEY on a tuple");
+    }
+    if (columns[0].sub) {
+        throw exceptions::unsupported_operation_exception("CONTAINS KEY lhs is subscripted");
+    }
+    if (!key) {
+        return true; // Compatible with old code, which skips null terms in key comparisons.
+    }
+    auto cdef = columns[0].col;
+    const auto collection = get_value(columns[0], data);
+    if (!collection) {
+        return false;
+    }
+    const auto data_map = value_cast<map_type_impl::native_type>(cdef->type->deserialize(*collection));
+    auto key_type = static_pointer_cast<const collection_type_impl>(cdef->type)->name_comparator();
+    auto found = with_linearized(*key, [&] (bytes_view k_bv) {
+        using entry = std::pair<data_value, data_value>;
+        return std::find_if(data_map.begin(), data_map.end(), [&] (const entry& element) {
+            return key_type->compare(element.first.serialize_nonnull(), k_bv) == 0;
+        });
+    });
+    return found != data_map.end();
+}
+
+/// Fetches the next cell value from iter and returns its value as bytes_opt if the cell is atomic;
+/// otherwise, returns nullopt.
```

Comment:
The description doesn't match the code; the code fetches the next atomic cell regardless of the column type (all cells are atomic; but non-atomic columns are multi-cell).

nullopt is returned when we run out of cells, not if we read a non-atomic column.

Reply:
We're not supposed to run out of cells here; that would mean selection was not obeyed in row fetching.  But the cell could have a null value in the row, and that's when nullopt is returned.  I'll fix the blurb, thanks for catching it.

Reply:
Fixed.